API, Core: Add geometry and geography types support

Kontinuation opened the pull request:
This adds 2 primitive types to `iceberg-api` and `iceberg-core` for supporting geospatial data types, partially implementing the iceberg geo spec:  https://github.com/apache/iceberg/pull/10981

The newly added primitive types are:

* `geometry(C)`: geometries with linear/planar edge interpolation
* `geography(C, A)`: geometries with non-linear edge interpolation, the algorithm for interpolating edges is parameterized by `A`.

szehon-ho reviewed (commented):
Hey , really sorry for delay, let's work on it this week.  Some early comments

Kontinuation commented on api/src/main/java/org/apache/iceberg/Geography.java:
```diff
@@ -0,0 +1,127 @@
+
+  private final Geometry geometry;
+
+  public Geography(Geometry geometry) {
```
We define Geography this way to reuse the data structures of coordinates and geospatial objects (Points, LineString, etc.), as well as WKB/WKT functions provided by JTS. Geography does not interoperate with Geometry, they are different classes but have the same data structure.

szehon-ho replied:
Can we have a base class, and two implementing classes then?  its confusing as it is.

Kontinuation replied:
I'll try removing Geometry and Geography from iceberg-api, and use `ByteBuffer` as the underlying Java class for these types.

Kontinuation commented on api/src/main/java/org/apache/iceberg/expressions/BoundLiteralPredicate.java:
```diff
@@ -88,11 +91,64 @@ public boolean test(T value) {
         return String.valueOf(value).startsWith((String) literal.value());
       case NOT_STARTS_WITH:
         return !String.valueOf(value).startsWith((String) literal.value());
+      case ST_INTERSECTS:
```
Predicates and pruning were removed from this PR.

szehon-ho replied:
Thank you

szehon-ho commented on api/src/main/java/org/apache/iceberg/expressions/Literal.java:
```diff
@@ -71,6 +73,14 @@ static Literal<BigDecimal> of(BigDecimal value) {
+    return new Literals.GeometryLiteral(value);
+  }
+
+  static Literal<Geography> of(Geography value) {
```
I guess here, let's specifically not leak the Geography in the API, we should just have a wrapper class as well.

szehon-ho commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +562,123 @@ public int hashCode() {
+    }
+
+    public static GeographyType of(String crs) {
+      return of(crs, DEFAULT_ALGORITHM);
```
another option is to allow null right?  and the code can default it , as per the spec?

Kontinuation commented on core/src/main/java/org/apache/iceberg/SchemaParser.java:
```diff
@@ -141,11 +143,34 @@ static void toJson(Types.MapType map, JsonGenerator generator) throws IOExceptio
+    if (primitive.typeId() == Type.TypeID.GEOMETRY) {
+      Types.GeometryType geometryType = (Types.GeometryType) primitive;
+      generator.writeStartObject();
+      generator.writeStringField("type", "geometry");
```
Replaced constant literals such as "type", "geometry", "geography", "crs" with constants.

Kontinuation commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -70,6 +75,20 @@ public static Type fromTypeName(String typeString) {
       return TYPES.get(lowerTypeString);
     }
 
+    if (lowerTypeString.startsWith("geometry")) {
```
We cannot match the regex with `lowerTypeString`, because we want to extract the original CRS from the type instead of a lower case one. I'll see how to make it more concise and get rid of startsWith.

rdblue replied:
So we consider CRS to be case sensitive?

The capture group of a case insensitive regular expression returns the matched characters without modification. So if CRS is case sensitive, you can still use a regular expression, like this:

```java
    Pattern pattern = Pattern.compile("geometry\\(([^)]*)\\)", Pattern.CASE_INSENSITIVE);
```

This test will pass:
```java
    Matcher m = pattern.matcher("GEOMETRY(aBc)");
    assertThat(m.matches()).isTrue();
    assertThat(m.group(1)).isEqualTo("aBc");
```

I'd prefer using a case insensitive pattern on `typeString` like this, rather than `startsWith`.

Kontinuation replied:
Switched to using case insensitive regex pattern matching.

Kontinuation commented on core/src/main/java/org/apache/iceberg/SchemaParser.java:
```diff
@@ -141,11 +143,34 @@ static void toJson(Types.MapType map, JsonGenerator generator) throws IOExceptio
 
   static void toJson(Type.PrimitiveType primitive, JsonGenerator generator) throws IOException {
-    generator.writeString(primitive.toString());
+    if (primitive.typeId() == Type.TypeID.GEOMETRY) {
```
Refactored to use switch-case.

Kontinuation commented on api/src/main/java/org/apache/iceberg/types/TypeUtil.java:
```diff
@@ -535,6 +535,8 @@ private static int estimateSize(Type type) {
       case BINARY:
       case VARIANT:
+      case GEOMETRY:
+      case GEOGRAPHY:
```
We cannot give an accurate estimation without statistics of the data, so I've simply reused the number for BINARY and VARIANT.

80 is roughly size of a polygon or linestring with 4 coordinates, which denotes a box. This is larger for a dataset full of points and smaller for datasets containing complex shapes, but I think it is a reasonable guess.

Kontinuation replied:
I've added reasons why we use 80 (even though the choice was made randomly like https://github.com/apache/iceberg/pull/11324/files#r1821277533) and avoid falling back into the same case as BINARY and VARIANT.

Kontinuation commented:
I have removed dependency to JTS from iceberg-api, now the underlying Java type of Geometry and Geography are `ByteBuffer`. Now JTS is a implementation dependency of iceberg-core.

I have removed code for converting between ByteBuffer and geospatial types, removed GeometryLiteral/GeographyLiteral, as well as default initial/write value support for geospatial fields. I'll proceed adding them once we are all agree on the API design.

hsiang-c commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -70,6 +74,20 @@ public static Type fromTypeName(String typeString) {
+      Matcher geography = GEOGRAPHY_PARAMETERS.matcher(typeString.substring(9));
+      if (geography.matches()) {
+        return GeographyType.of(geography.group(1), geography.group(2));
+      }
```
In both `geometry` and `geography` cases, do you think we should throw `IllegalArgumentException` when patterns don't match?

Kontinuation replied:
We'll throw an `IllegalArgumentException` at the bottom of this function if none of the pattern matches: `"Cannot parse type string to primitive: " + typeString`. I think this is already informative. Do you think that we need a more specific error message for geospatial types?

hsiang-c replied:
No worries, I didn't pay attention to the very bottom. LGTM.

hsiang-c commented on api/src/main/java/org/apache/iceberg/transforms/Identity.java:
```diff
@@ -19,16 +19,21 @@
 class Identity<T> implements Transform<T, T> {
   private static final Identity<?> INSTANCE = new Identity<>();
 
+  private static final Set<Type.TypeID> UNSUPPORTED_TYPES =
```
👍

hsiang-c commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +561,120 @@ public int hashCode() {
+    }
+
+    public static GeometryType of(String crs) {
+      return new GeometryType(crs == null ? "" : crs);
```
(nit) Maybe `crs == null` can be `crs == null || crs.isEmpty()`

Kontinuation replied:
I believed that including crs.isEmpty() was redundant in this situation, so I didn't add it, as we would return an empty string regardless.

rdblue replied:
The empty string should not be allowed. That is an invalid CRS because it has no information. Null is fine to support here if you expect to have situations in which callers will find it convenient to pass null to mean "use the default".

hsiang-c commented on core/src/main/java/org/apache/iceberg/util/GeometryUtil.java:
```diff
@@ -0,0 +1,75 @@
+    // We need to set outputDimension = 4 for XYM geometries to make JTS WKTWriter or WKBWriter work
+    // correctly.
+    // The WKB/WKT writers will ignore Z ordinate for XYM geometries.
+    if (!Double.isNaN(coordinate.getZ())) {
```
Perhaps we could do (not a must but more readable to me).

```java
if (Coordinate.NULL_ORDINATE != coordinate.getZ())
```

Kontinuation replied:
I'm afraid that we cannot do this. `Coordinate. NULL_ORDINATE` is NaN, comparing it with other floating point number using != will always yield true.

hsiang-c commented on core/src/test/java/org/apache/iceberg/util/TestGeometryUtil.java:
```diff
@@ -0,0 +1,145 @@
+import org.locationtech.jts.geom.Geometry;
+import org.locationtech.jts.geom.GeometryFactory;
+
+public class TestGeometryUtil {
```
It seems like you covered `Coordinate` and its subclasses except `ExtendedCoordinate`, am I right?

Kontinuation replied:
Yes. As far as I know, `ExtendedCoordinate` is part of JTS example code so we don't need to handle it. JTS will only be used internally in iceberg and it exchanges geospatial data with other systems through WKB/Bounds byte buffer, so there's no need to take non-standard extensions of `Coordinate` into consideration.

hsiang-c replied:
Thanks for the explanation.

hsiang-c reviewed (approved):
LGTM

szehon-ho commented on api/src/main/java/org/apache/iceberg/types/EdgeInterpolationAlgorithm.java:
```diff
@@ -0,0 +1,64 @@
+  }
+
+  public static EdgeInterpolationAlgorithm fromName(String algorithmName) {
+    try {
```
Precondition, check null and throw exception?

szehon-ho commented on api/src/main/java/org/apache/iceberg/types/EdgeInterpolationAlgorithm.java:
```diff
@@ -0,0 +1,64 @@
+      return EdgeInterpolationAlgorithm.valueOf(algorithmName.toUpperCase(Locale.ENGLISH));
+    } catch (IllegalArgumentException e) {
+      throw new IllegalArgumentException(
+          String.format("Invalid edge interpolation algorithm name: %s", algorithmName), e);
```
Nit: can remove redundant 'name'

szehon-ho commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +561,120 @@ public int hashCode() {
+    }
+
+    public static GeometryType get() {
+      return of("");
```
why not , new GeometryType("")

Kontinuation replied:
Fixed. Now we are using `null` as default value as requested. For now it is `new GeometryType(null)`.

szehon-ho commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +561,120 @@ public int hashCode() {
+    }
+
+    public static GeographyType get() {
+      return of("");
```
why not new GeographyType("")

Kontinuation replied:
Fixed. Now it is `new GeographyType(null, null)`.

szehon-ho commented on api/src/test/java/org/apache/iceberg/types/TestTypes.java:
```diff
@@ -90,5 +90,42 @@ public void testNestedFieldBuilderIdCheck() {
         .isThrownBy(() -> required("field").ofType(Types.StringType.get()).build())
         .withMessage("Id cannot be null");
+
+    assertThat(Types.fromPrimitiveString("geometry")).isEqualTo(Types.GeometryType.get());
```
this should be in its own test

rdblue replied:
In this suite, these cases would be added to `fromPrimitiveString` and `fromTypeName`.

Kontinuation replied:
Moved tests to `fromTypeName` and `fromPrimitiveString`.

szehon-ho commented on core/src/main/java/org/apache/iceberg/SchemaParser.java:
```diff
@@ -277,6 +313,17 @@ private static Types.MapType mapFromJson(JsonNode json) {
   }
 
+  private static Types.GeometryType geometryFromJson(JsonNode json) {
+    String crs = JsonUtil.getStringOrNull("crs", json);
```
nit: these (and following) can be replaced by the constants?

Kontinuation replied:
This is reverted since we are not encoding geospatial types as JSON.

szehon-ho commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +561,120 @@ public int hashCode() {
+      return TypeID.GEOMETRY;
+    }
+
+    public String crs() {
```
maybe can annotate NotNull?

Kontinuation replied:
The `NotNull` annotation requires an additional dependency, and I noticed that it is not currently used in iceberg-api or iceberg-core. Would you like me to add the additional dependency and annotate this method?

szehon-ho replied:
Got it, nvm it may not be worth it

rdblue replied:
Iceberg does not generally use nullability annotations.

szehon-ho commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +561,120 @@ public int hashCode() {
+    private final String crs;
+
+    private GeometryType(String crs) {
+      this.crs = crs;
```
Precondition that crs is not null?

rdblue replied:
I think it would be easier to allow null to indicate the default. That's easier to detect for `toString` to show `geometry` instead of `geometry()`.

Kontinuation replied:
We allow `null` and not allow empty string for CRS. Additionally we check that the CRS should not have leading or trailing spaces to make sure that the value of CRS will be persisted after converting to string and parsing it back.

szehon-ho replied:
Curious, why not always have toString show the default value like Geography?  Geometry("OGC:CRS84").  Is it to save some bytes ?  Seems to be easier imo.

Kontinuation replied:
Geography does not always show the default CRS. It happens when the edge algorithm is not using default value but CRS is left unset. We have to fill in the default CRS as the first argument in this case.

I thought that the behavior of toString is to omit optional parameters when possible. See https://github.com/apache/iceberg/pull/12346/files/0c451cac53f4098680aa0ae518af70dc45efaeb1#r1994430054.

szehon-ho commented on core/src/main/java/org/apache/iceberg/util/GeometryUtil.java:
```diff
@@ -0,0 +1,75 @@
+  }
+
+  private static int getOutputDimension(Geometry geom) {
+    int dimension = 2;
```
nit: make a constant to avoid instantiating it every time?

Kontinuation replied:
This file is removed as we don't need WKB/WKT serialization according to the latest spec.

szehon-ho reviewed (commented):
Thanks @Kontinuation for removing the api dependency on jts from iceberg-api, this looks a lot closer to me

szehon-ho commented:
Also, (as can't comment on files that are not in the change)

Do we need to add Geo types to following places?

1.  Types.java TYPES constant?
2. TestSchemaUnionByFieldName primitiveTypes()?
3. TestSchema TEST_TYPES?
4. TestTypesUtil testTypes()?

rdblue commented on api/src/main/java/org/apache/iceberg/transforms/Identity.java:
```diff
@@ -93,6 +98,9 @@ public SerializableFunction<T, T> bind(Type type) {
   public boolean canTransform(Type maybePrimitive) {
+    if (UNSUPPORTED_TYPES.contains(maybePrimitive.typeId())) {
+      return false;
+    }
```
Style: newlines are added after control flow blocks before the next statement.

rdblue commented on api/src/main/java/org/apache/iceberg/types/EdgeInterpolationAlgorithm.java:
```diff
@@ -0,0 +1,66 @@
+/** The algorithm for interpolating edges. */
+public enum EdgeInterpolationAlgorithm {
+  /** Edges are interpolated as geodesics on a sphere. */
+  SPHERICAL("spherical"),
```
Minor: We usually only include the lower-case version when it isn't a simple translation from the name.

rdblue commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -58,9 +58,15 @@ private Types() {}
+  private static final Pattern GEOMETRY_PARAMETERS =
+      Pattern.compile("(?:\\(\\s*([^, ]+)?\\s*\\))?");
+  private static final Pattern GEOGRAPHY_PARAMETERS =
+      Pattern.compile("(?:\\(\\s*([^, ]+)?\\s*(?:,\\s*(\\w*)\\s*)?\\))?");
```
I think the intent with the first capture group is to avoid consuming trailing spaces between the group and the comma, but the current pattern would fail when there are spaces in the CRS name, like `EPSG: 4326`. We may want to support names with spaces because they are unambiguous, and we don't have anything that disallows spaces in those names in the specs.

Instead of matching non-space characters, I recommend matching any character other than comma using a non-greedy `+` by adding `?`: `[^,]+?`

The full regex is `Pattern.compile("(?:\\(\\s*([^,]+?)?\\s*(?:,\\s*(\\w*)\\s*)?\\))?");` and that works:

```java
    Pattern p2 = Pattern.compile("(?:\\(\\s*([^,]+?)?\\s*(?:,\\s*(\\w*)\\s*)?\\))?");
    Matcher m2 = p2.matcher("(EPSG: 4326  , vincenty)");
    assertThat(m2.matches()).isTrue();
    assertThat(m2.group(1)).isEqualTo("EPSG: 4326");
    assertThat(m2.group(2)).isEqualTo("vincenty");
```

Kontinuation replied:
Fixed. Now we allow spaces inside CRS.

rdblue commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -70,6 +76,20 @@ public static Type fromTypeName(String typeString) {
+    if (lowerTypeString.startsWith("geography")) {
+      Matcher geography = GEOGRAPHY_PARAMETERS.matcher(typeString.substring(9));
+      if (geography.matches()) {
+        return GeographyType.of(geography.group(1), geography.group(2));
```
I think that this should validate the algorithm and pass the algorithm enum symbol instead of a string. That avoids being able to compile with a typo (because `EdgeInterpolationAlgorithm.VINCENTI` is not a symbol) instead of allowing code paths that contain `"vincenti"`.

Kontinuation replied:
Validated algorithm here.

rdblue commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +563,122 @@ public int hashCode() {
+    private final String crs;
+
+    private GeometryType(String crs) {
+      Preconditions.checkNotNull(crs, "CRS cannot be null");
```
Style: In Iceberg, we try to have direct and clear error messages by cutting out sentence structure. Here, we would normally use `"Invalid CRS: null"`.

Kontinuation replied:
We allow `null` but checks for empty string and leading/trailing spaces. The error messages are:

* `Invalid CRS: (empty string)` for empty string
* `CRS must not have leading or trailing spaces: '%s'` for CRS with leading or trailing spaces.

rdblue commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +563,122 @@ public int hashCode() {
+    }
+
+    public static GeometryType get() {
+      return new GeometryType("");
```
Please use a valid CRS rather than an empty string. I think that this defaults to CRS84, which seems like a reasonable string to use.

Kontinuation replied:
We are using `null` for now. Now I believe that using `OGC:CRS84` could be a better choice. See the comments for `GeographyType.toString` below.

rdblue replied:
I think `null` is a good choice for the default.

rdblue commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +563,122 @@ public int hashCode() {
+
+    @Override
+    public String toString() {
+      return String.format("geometry(%s)", crs);
```
Using empty string for CRS would produce `geometry()` instead of `geometry`. I would also not use `geometry(crs84)` when that is equivalent to `geometry`. So whatever we store for the default CRS (null?) this should translate the default to `geometry`.

Kontinuation replied:
The default CRS is stored as `null`, it will be translated to `geometry` for such cases.

rdblue replied:
Looks good. The only remaining case is normalizing `OGC:CRS84`, which I suggested in the constructor comment above.

rdblue commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +563,122 @@ public int hashCode() {
+    }
+
+    public static GeographyType get() {
+      return new GeographyType("", null);
```
Similar to the comments above: the method to get a default type should not be `get` and an empty string should not be used to carry information. Use either `null` or `CRS84`. I prefer `null`.

Kontinuation replied:
Renamed this method to `crs84()`. Use `null` for default CRS.

rdblue commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +563,122 @@ public int hashCode() {
+      return new GeographyType("", null);
+    }
+
+    public static GeographyType of(String crs) {
```
Is there a better name than `of` that is more descriptive? I think that `forCRS` would work here, but it is a bit confusing when the algorithm is added (`forCRS("srid:1234", EdgeInterpolationAlgorithm:ANDOYER)`. Maybe `of` is the best choice.

Kontinuation replied:
I just realized that we should be careful about leading and trailing spaces in `crs`. Should we trim the `crs` before constructing geospatial type objects or throw an exception when leading/trailing spaces are found in `crs`?

Kontinuation replied:
I decided to throw exception when crs has leading/trailing spaces.

szehon-ho replied:
if we have two methods (the other one taking crs/algorithm), why not rename this to forCRS now?

Kontinuation replied:
Renamed to `forCRS`.

rdblue commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +563,122 @@ public int hashCode() {
+      return new GeographyType(crs, algorithm);
+    }
+
+    public static GeographyType of(String crs, String algorithmName) {
```
I think this should be removed from the API and the logic should be moved into the parsing code.

Kontinuation replied:
Removed this API.

rdblue commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +563,122 @@ public int hashCode() {
+
+    @Override
+    public String toString() {
+      return String.format("geography(%s, %s)", crs, algorithm != null ? algorithm.value() : "");
```
Looks like this will produce `"geography(, )"` for the default type. When the type is the default, it should produce just `geography`. When algorithm is not set, it should produce `geography(crs)` and when algorithm is set it should produce `geography(crs, algorithm)`. When the CRS is defaulted and algorithm is not, it should use whatever standard name we decide on for CRS84.

Kontinuation replied:
The problem is that there's no standard name for CRS84. `OGC:CRS84` mentioned in the spec is not a valid value for CRS. I suggest that we change the spec to allow `OGC:CRS84` to represent the default CRS. @jiayuasu @szehon-ho

szehon-ho replied:
I think the spec says 'ogc:crs84' is default, and 'srid' and 'projjson' is for extensions, so its ok as a default value for me.  Wdyt @jiayuasu ?

Kontinuation replied:
I am using `OGC:CRS84` for CRS when only algorithm is specified. One consequence is that when the serialized type is parsed back to `GeographyType` object, the `crs` field will be `OGC:CRS84` instead of `null`. I'm not sure if we could leave such behavior as is.

rdblue commented on api/src/test/java/org/apache/iceberg/TestPartitionSpecValidation.java:
```diff
@@ -327,12 +329,34 @@ public void testVariantUnsupported() {
   }
 
+  @Test
+  public void testGeometryUnsupported() {
```
Can these be collapsed with `testVariantUnsupported` in a `ParameterizedTest`?

rdblue replied:
Can you also make sure that `identity` is tested like this?

Kontinuation replied:
Refactored them as `TestIdentity.testUnsupported` and `TestPartitionSpecValidation.testUnsupported`.

rdblue commented on api/src/main/java/org/apache/iceberg/types/EdgeInterpolationAlgorithm.java:
```diff
@@ -0,0 +1,66 @@
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+
+/** The algorithm for interpolating edges. */
+public enum EdgeInterpolationAlgorithm {
```
This is a long name to use as a prefix. What about `EdgeAlgorithm` instead?

Also, I'm not sure that "interpolation" is the right word for this because the algorithm calculates points on the edge but does not define the edge. The CRS defines the interpolation for the edge, right?

Kontinuation replied:
I've renamed it to `EdgeAlgorithm` in the last commit, as the section name for the definitions of these algorithms is Edge-Interpolation Algorithm.

rdblue commented on api/src/test/java/org/apache/iceberg/types/TestSerializableTypes.java:
```diff
@@ -63,7 +63,14 @@ public void testEqualTypes() throws Exception {
+          Types.FixedType.ofLength(34),
+          Types.GeometryType.get(),
+          Types.GeometryType.of("srid:3857"),
+          Types.GeometryType.of("projjson:Test_Identifier"),
```
Is this valuable? These tests are the same for any string.

Kontinuation replied:
PrimitiveTypes are serialized and deserialized using `toString` and `fromTypeString`, since PrimitiveType has implemented `Object writeReplace()` method to handle Java serialization. Type parameters do affect how primitive types are serialized and deserialized.

A better place for such tests could be `TestTypes.java`, but I think that having tests here to ensure round-trip type consistency also makes sense.

rdblue replied:
Sorry, I meant is it valuable to have a `srid` case and a `projjson` case. Both are strings and if we validate that any string is preserved by serialization and deserialization, I think that is sufficient.

Kontinuation replied:
I'm removing the case with projjson and only preserves the one with srid.

rdblue commented on build.gradle:
```diff
@@ -348,6 +348,7 @@ project(':iceberg-core') {
     implementation libs.jackson.databind
     implementation libs.caffeine
     implementation libs.roaringbitmap
+    implementation libs.jts.core
```
Why is this needed?

rdblue replied:
I see, this is used for translation between WKT and WKB.

If we choose to do that translation (we could change from a WKT representation or could not allow defaults for geo fields) then we will need to update license documentation or make support pluggable. Updating documentation involves researching the code that is pulled in by JTS and reviewing the license and notice for every new transitive dependency.

Kontinuation replied:
We only allow `null` defaults for geospatial types, so we don't need to implement JSON single value parsing for geospatial types. I have removed single value parsing support for geospatial types and also removed JTS.

rdblue commented on core/src/main/java/org/apache/iceberg/SchemaParser.java:
```diff
@@ -141,11 +145,39 @@ static void toJson(Types.MapType map, JsonGenerator generator) throws IOExceptio
 
   static void toJson(Type.PrimitiveType primitive, JsonGenerator generator) throws IOException {
-    generator.writeString(primitive.toString());
+    switch (primitive.typeId()) {
```
Primitives are encoded as type strings. This method should not change and `GeometryType.toString` should produce a string that can be parsed by `Types.fromTypeName`.

Kontinuation replied:
Geography and Geometry types are encoded as JSON objects according to https://iceberg.apache.org/spec/#schemas, we can update the spec if strings are always preferred for primitive types including geospatial types.

rdblue replied:
I missed that in the spec.

@szehon-ho, why are we using objects for these when we also support parsing them as we do for `decimal` and `fixed`? Should we update the spec?

szehon-ho replied:
Im ok with that, initially I thought its a bit complex as the crs can be projjson content.  But now it is a table property, so should work.

rdblue commented on core/src/main/java/org/apache/iceberg/SchemaParser.java:
```diff
@@ -141,11 +145,39 @@ static void toJson(Types.MapType map, JsonGenerator generator) throws IOExceptio
-    if (type.isPrimitiveType() || type.isVariantType()) {
+    if (type.isPrimitiveType()) {
+      toJson(type.asPrimitiveType(), generator);
+    } else if (type.isVariantType()) {
```
This should not change.

Kontinuation replied:
Reverted according to latest spec change. Geospatial types are represented by strings rather than JSON objects.

rdblue commented on core/src/main/java/org/apache/iceberg/SingleValueParser.java:
```diff
@@ -335,6 +349,13 @@ public static void toJson(Type type, Object defaultValue, JsonGenerator generato
         }
         break;
+      case GEOMETRY:
+      case GEOGRAPHY:
```
@szehon-ho, is the WKT representation worth adding a library to translate between WKT and WKB? Maybe this should do the same thing as binary like we do for variant?

Kontinuation replied:
The JSON single-value serialization for geospatial types could only be used for initial/write default values. We found no real-world scenarios that require defining custom default values other than `null` for geospatial fields, so this WKT representation is not being used. We can leave JSON single-value serialization for geospatial values unimplemented and also remove JTS from iceberg dependency.

rdblue replied:
If there aren't real-world scenarios for default values, then we should update the spec to ensure that the default is always null.

szehon-ho replied:
Noted, will take a look at this

rdblue commented on core/src/main/java/org/apache/iceberg/util/GeometryUtil.java:
```diff
@@ -0,0 +1,77 @@
+    }
+  }
+
+  private static int getOutputDimension(Geometry geom) {
```
Style: Even for private methods, we avoid using `get` in method names. It is either not adding value (can be omitted) or should be replaced with a more specific verb. I would probably omit here and name this `dimension(Geometry)`.

Kontinuation replied:
We removed this file, as we removed JTS and single value serialization support for geospatial types.

rdblue commented on core/src/main/java/org/apache/iceberg/util/GeometryUtil.java:
```diff
@@ -0,0 +1,77 @@
+    if (!Double.isNaN(coordinate.getZ())) {
+      dimension = 3;
+    }
+    if (!Double.isNaN(coordinate.getM())) {
```
Style: empty lines following control flow blocks.

Kontinuation replied:
We removed this file, as we removed JTS and single value serialization support for geospatial types.

rdblue commented on core/src/test/java/org/apache/iceberg/TestGeospatialTable.java:
```diff
@@ -0,0 +1,69 @@
+import org.apache.iceberg.types.Types;
+import org.junit.jupiter.api.Test;
+
+public class TestGeospatialTable extends HadoopTableTestBase {
```
Why does this use `HadoopTableTestBase`? I would avoid using `HadoopCatalog` or considering something that works with `HadoopCatalog` to be working everywhere.

Kontinuation replied:
Moved this test to `CatalogTests`.

Kontinuation replied:
Yes, it validate catalog behavior. By moving it to CatalogTest we'll be able to test it against all catalog implementations. We found that HiveCatalog does not support geospatial types and fixed it in the previous commit. If you insist that we only test using `InMemoryCatalog`, I'll bring `TestGeospatialTable` back and test using `InMemoryCatalog` only.

rdblue commented on core/src/test/java/org/apache/iceberg/TestSchemaParser.java:
```diff
@@ -142,4 +143,31 @@ public void testPrimitiveTypeDefaultValues(Type.PrimitiveType type, Literal<?> d
   }
+
+  @Test
+  public void testVariantType() throws IOException {
```
This is tested because `supportsVariantType` at the top is true. I would use the same pattern for spatial types, adding them to the list in `DataTest` and disabling the tests with a `supports` method. That way we can slowly add support and enable the tests (as we are doing for unknown, variant, and timestamp(9)).

Kontinuation replied:
Removed `testVariantType` and updated `DataTest` to testing geospatial types. `supportsGeospatial` is added to `DataTest` and default to `false`. It is enabled in `TestSchemaParser`.

rdblue commented on core/src/test/java/org/apache/iceberg/TestSchemaUpdate.java:
```diff
@@ -364,7 +365,11 @@ public void testUpdateFailure() {
+            Types.GeometryType.get(),
+            Types.GeometryType.of("srid:3857"),
+            Types.GeographyType.get(),
+            Types.GeographyType.of("srid:4269", EdgeInterpolationAlgorithm.KARNEY));
```
Good catch with this test. @aihuaxu we will need to update this for variant and other types as well.

rdblue commented on core/src/test/java/org/apache/iceberg/TestSingleValueParser.java:
```diff
@@ -157,6 +167,15 @@ public void testInvalidTimestamptz() {
+    String defaultJson = "\"POINT (1 2 3 4 5 6)\"";
+    assertThatThrownBy(() -> defaultValueParseAndUnParseRoundTrip(expectedType, defaultJson))
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageMatching("Cannot parse default as a geometry.* value.*");
```
What are the `.*` patterns matching?

Kontinuation replied:
Geometry types without CRS were serialized as `geometry()`. The last commit fixed this so `.*` pattern was removed.

Kontinuation commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +563,122 @@ public int hashCode() {
+      this.crs = crs;
+    }
+
+    public static GeometryType get() {
```
Renamed get() methods of GeometryType and GeographyType to `crs84()`.

rdblue commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -70,6 +76,25 @@ public static Type fromTypeName(String typeString) {
+      if (geography.matches()) {
+        String algorithmName = geography.group(2);
+        EdgeAlgorithm algorithm =
+            (algorithmName == null || algorithmName.isEmpty())
```
Why is `algorithmName.isEmpty` valid here? I think it should be an error if the type is passed as `"geography(srid:1234,)"`. If there is a non-null string then it must be a valid algorithm name. When there is no comma, the second capture group is null, so removing this `isEmpty` check will fix it.

Kontinuation replied:
Removed `isEmpty`.

rdblue commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +568,124 @@ public int hashCode() {
+      if (algorithm != null) {
+        return String.format(
+            "geography(%s, %s)",
+            crs != null ? crs : "", algorithm.name().toLowerCase(Locale.ENGLISH));
```
I think it would be fine to move the `toLowerCase` into the enum. We just don't need to parameterize the enum with a constant that is the lower case name.

Kontinuation replied:
Implemented `toString` in the enum to perform `toLowerCase`.

rdblue commented on api/src/test/java/org/apache/iceberg/types/TestTypes.java:
```diff
@@ -81,6 +81,56 @@ public void fromPrimitiveString() {
+  @Test
+  public void geospatialTypeFromTypeName() {
+    assertThat(Types.fromPrimitiveString("geometry")).isEqualTo(Types.GeometryType.crs84());
+    assertThat(Types.fromPrimitiveString("geometry()")).isEqualTo(Types.GeometryType.crs84());
```
Why is this a valid? It seems like a problem to me. We don't allow `long()`.

Kontinuation replied:
Fixed. `geometry()` and `geography()` are not valid representations of geospatial types.

rdblue commented on core/src/test/java/org/apache/iceberg/TestTableMetadata.java:
```diff
@@ -1825,6 +1825,43 @@ public void testConstructV3Metadata() {
   }
 
+  @Test
+  public void testV3GeometryTypeSupport() {
```
I think a better place for this to be tested in in `TestSchema.testUnsupportedTypes`. There are already a good set of cases for nesting and other types.

Kontinuation replied:
We've already updated `TestSchema` to make use of `testUnsupportedTypes` and `testTypeSupported` to test geospatial types. I'm removing this test case from `TestTableMetadata`.

rdblue commented on core/src/test/java/org/apache/iceberg/data/DataTest.java:
```diff
@@ -129,28 +135,37 @@ protected boolean supportsVariant() {
     writeAndValidate(new Schema(required(1, "id", LongType.get()), optional(2, "test_type", type)));
   }
 
+  private static void assumeNoUnsupportedType(
```
I'd rather not refactor this out since it is a method added with just one statement.

szehon-ho commented:
@rdblue @Kontinuation i updated the spec here with the minor suggested fixes: https://github.com/apache/iceberg/pull/12533

Kontinuation commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +568,124 @@ public int hashCode() {
+
+    @Override
+    public String toString() {
+      if (Strings.isNullOrEmpty(crs)) {
```
Fixed. Now we only check for `null`.

Kontinuation commented:
> Also, (as can't comment on files that are not in the change)
> 
> Do we need to add Geo types to following places?
> 
> 1. Types.java TYPES constant?
> 2. TestSchemaUnionByFieldName primitiveTypes()?
> 3. TestSchema TEST_TYPES?
> 4. TestTypesUtil testTypes()?

Tests were added to the above mentioned places.

szehon-ho commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +565,134 @@ public int hashCode() {
+    private GeometryType(String crs) {
+      if (crs != null) {
+        Preconditions.checkArgument(!crs.isEmpty(), "Invalid CRS: (empty string)");
+        Preconditions.checkArgument(
```
i might have missed this, but why not just trim it?

Kontinuation replied:
Constructing a GeometryType manually using `GeometryType.of(" crs_value  ")` and retrieve its CRS will result in a different value if we trim it. I've asked for it here https://github.com/apache/iceberg/pull/12346#discussion_r1996626058 and decided to throw exception to avoid any unexpected behavior for such cases. I can change it to silently trimming the input if you consider it a more appropriate approach.

rdblue replied:
I don't think this check is needed. We don't do similar validations elsewhere. The responsibility of this class is to pass the incoming CRS without modification. There's no need to validate the string. Parsing should handle that.

Kontinuation replied:
Removed trim. Trimming is handled by regex when parsing primitive type strings.

szehon-ho commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +565,134 @@ public int hashCode() {
+
+    private GeometryType(String crs) {
+      if (crs != null) {
+        Preconditions.checkArgument(!crs.isEmpty(), "Invalid CRS: (empty string)");
```
why the extra parens around 'empty string'?

Kontinuation replied:
This is following the existing error message style: https://github.com/apache/iceberg/blob/apache-iceberg-1.8.1/core/src/main/java/org/apache/iceberg/view/BaseView.java#L116, I thought that it is a convention to add parenthesis around empty string.

szehon-ho commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +565,134 @@ public int hashCode() {
+    }
+
+    public static GeometryType crs84() {
+      return new GeometryType(null);
```
style: what do you think to have a private GeometryType() constructor that does nothing?  to avoid having null passed in explicitly

Kontinuation replied:
Added a new zero-arg constructor. Similar change was made for GeographyType.

szehon-ho commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +565,134 @@ public int hashCode() {
+        Preconditions.checkArgument(
+            crs.trim().equals(crs), "CRS must not have leading or trailing spaces: '%s'", crs);
+      }
+      this.crs = crs;
```
nit: put this in the if block (same for geometry?)

Kontinuation replied:
I have moved `crs` to the if block. I also have to add an else block to assign `crs` as null because `crs` is final.

szehon-ho commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +565,134 @@ public int hashCode() {
+
+    private GeographyType(String crs, EdgeAlgorithm algorithm) {
+      if (crs != null) {
+        Preconditions.checkArgument(!crs.isEmpty(), "Invalid CRS: (empty string)");
```
same comment, i would remove the parens.

rdblue commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -58,9 +58,16 @@ private Types() {}
+      Pattern.compile("geometry\\s*(?:\\(\\s*([^,]+?)\\s*\\))?", Pattern.CASE_INSENSITIVE);
+  private static final Pattern GEOGRAPHY_PARAMETERS =
+      Pattern.compile(
+          "geography\\s*(?:\\(\\s*([^,]+)\\s*(?:,\\s*(\\w*)\\s*)?\\))?", Pattern.CASE_INSENSITIVE);
```
This doesn't use non-greedy matching for the CRS, which means that the `trim` call below is required. I think it would be better to use non-greedy matching using `([^,]+?)` and remove the call to `trim`.

Kontinuation replied:
I have tweaked the regex to make it work without requiring trimming,

rdblue commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -58,9 +58,16 @@ private Types() {}
 
   private static final Pattern FIXED = Pattern.compile("fixed\\[\\s*(\\d+)\\s*\\]");
+  private static final Pattern GEOMETRY_PARAMETERS =
+      Pattern.compile("geometry\\s*(?:\\(\\s*([^,]+?)\\s*\\))?", Pattern.CASE_INSENSITIVE);
```
I typically like to match on any character other than the next delimiter, which in this case is `)`. Rather than matching any non-comma character with `([^,]+?)`, using `([^)]+?)` instead will cover more cases.

We don't want a string to fail matching the regex but that's what would happen in cases like `geometry(srid:123,456)`. Instead of failing to match, I think it will be better to parse correctly and then state that the CRS is unknown.

Kontinuation replied:
Use `([^)]+?)` to capture more cases and check for `,` after matching.

rdblue commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -70,6 +77,21 @@ public static Type fromTypeName(String typeString) {
+      String algorithmName = geography.group(2);
+      EdgeAlgorithm algorithm =
+          algorithmName == null ? null : EdgeAlgorithm.fromName(algorithmName.trim());
+      return GeographyType.of(crs != null ? crs.trim() : null, algorithm);
```
Are calls to `trim` needed?

Kontinuation replied:
Removed trim, as the regex was modified to handle trimming.

rdblue commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +565,148 @@ public int hashCode() {
+      crs = null;
+    }
+
+    public static GeometryType crs84() {
```
Static factory methods typically go above constructors, and the no-arg constructor usually comes first.

Kontinuation replied:
Reordered.

* Static members comes first
* Then static factory methods
* Then private members
* Then no-arg constructor
* Then other constructors

rdblue commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +565,148 @@ public int hashCode() {
+
+    private final String crs;
+
+    private GeometryType(String crs) {
```
I think this can be simpler:

```java
    private GeometryType(String crs) {
      Preconditions.checkArgument(crs == null || !crs.isEmpty(), "Invalid CRS: (empty string)");
      this.crs = CRS84.equalsIgnoreCase(crs) ? null : crs;
```

Kontinuation replied:
Use this simplified version.

rdblue commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +565,148 @@ public int hashCode() {
+      return new GeographyType();
+    }
+
+    public static GeographyType forCRS(String crs) {
```
I think I disagree with @szehon-ho for the name here. I think that `of` is fine because the other version of this method accepts a CRS string and `EdgeAlgorithm`. I'm mainly motivated here by making this method match what we use for `geometry`.

szehon-ho replied:
Sounds good

Kontinuation replied:
Renamed it back to `of`.

rdblue commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +565,148 @@ public int hashCode() {
+      this.algorithm = null;
+    }
+
+    public static GeographyType crs84() {
```
Could you move static factory methods to the top?

Kontinuation replied:
Reordered static factory methods.

rdblue commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +565,148 @@ public int hashCode() {
+            crs.trim().equals(crs), "CRS must not have leading or trailing spaces: '%s'", crs);
+        this.crs = crs;
+      } else {
+        this.crs = null;
```
Same comments as above for making this method simpler and normalizing the CRS when it is `OGC:CRS84`.

Kontinuation replied:
Changed it to be similar to the constructor of `GeometryType`.

rdblue commented on api/src/test/java/org/apache/iceberg/types/TestTypes.java:
```diff
@@ -79,6 +95,77 @@ public void fromPrimitiveString() {
+
+    assertThatExceptionOfType(IllegalArgumentException.class)
+        .isThrownBy(() -> Types.fromPrimitiveString("geometry()"))
+        .withMessageContaining("Cannot parse type string to primitive");
```
@szehon-ho what do you think the behavior here should be? I would expect it to match the following case and throw `Invalid CRS: (empty string)`.

szehon-ho replied:
Yea I think so, the current message is not that clear.

Kontinuation replied:
I've tweaked the regex to match this regex with the first group being empty, now it will emit `Invalid CRS: (empty string)`.

rdblue commented on core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java:
```diff
@@ -536,6 +538,45 @@ public void testBasicCreateTable() {
   }
 
+  @Test
+  public void testCreateGeospatialTable() {
```
I think I commented elsewhere, but this is not a good place for an end-to-end test creating a table with a geospatial column. These tests assert the behavior of catalogs.

Kontinuation replied:
Moved it to `TestGeospatialTable` and switched from `HadoopCatalog` to `InMemoryCatalog`.

rdblue commented on core/src/test/java/org/apache/iceberg/data/DataTest.java:
```diff
@@ -129,24 +135,37 @@ protected boolean supportsVariant() {
-                || TypeUtil.find(type, t -> t.typeId() == Type.TypeID.VARIANT) == null)
-        .as("variant is not yet implemented")
-        .isTrue();
+    if (!supportsUnknown()) {
```
Can you revert the changes to the other assumptions? I don't think that we need to modify those lines.

Kontinuation replied:
Reverted

rdblue commented on hive-metastore/src/main/java/org/apache/iceberg/hive/HiveSchemaUtil.java:
```diff
@@ -166,6 +166,8 @@ private static String convertToTypeString(Type type) {
       case FIXED:
       case BINARY:
+      case GEOMETRY:
+      case GEOGRAPHY:
```
Why is this included? I don't think that we want to translate these into Hive types.

Kontinuation replied:
I've reverted this change. Does it mean that Hive catalog won't support geospatial columns?

rdblue replied:
When this is needed, we can add it. But there's no value in translating to binary right now.

rdblue commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -70,6 +77,22 @@ public static Type fromTypeName(String typeString) {
+    Matcher geometry = GEOMETRY_PARAMETERS.matcher(typeString);
+    if (geometry.matches()) {
+      String crs = geometry.group(1);
+      Preconditions.checkArgument(!crs.contains(","), "Invalid CRS: %s", crs);
```
Why is this invalid? This should pass the string as-is. It is up to others to determine what is a valid CRS.

Kontinuation replied:
I'll remove this check and pass the CRS string as-is, since it won't cause ambiguity for geometry types.

rdblue commented on api/src/main/java/org/apache/iceberg/types/Types.java:
```diff
@@ -543,6 +566,135 @@ public int hashCode() {
+    }
+
+    public String crs() {
+      return crs;
```
If CRS is null, should this return `DEFAULT_CRS`?

Kontinuation replied:
I think it is better to return `DEFAULT_CRS`, so that the callers don't have to handle both nulls and default CRS.

We can make `GeographyType.algorithm` return `EdgeAlgorithm.SPHERICAL` when algorithm is null as well.

rdblue commented:
Looks great! There are some minor nits, but I think it makes sense to commit this to move forward. Thanks @Kontinuation and @szehon-ho!

szehon-ho commented:
Thanks , great work @Kontinuation , and thanks @rdblue !
